Feature: remote browser support

adamhancock opened the pull request:
- [X] I have read and understand the pull request rules.

# Description

Adds Remote browser support. Similar to docker hosts with the ability to add multiple remote browser playwright configurations. Direct support to use with the docker image: https://hub.docker.com/r/browserless/chrome. 

Browserless Playwright docs: https://github.com/browserless/chrome#playwright

Related:

- https://github.com/louislam/uptime-kuma/pull/3308

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I ran ESLint and other linters for modified files
- [X] I have performed a self-review of my own code and tested it
- [X] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)
- [X] My changes generate no new warnings

louislam commented:
Good idea, it's great.

I am focusing on 2.0.0 right now. This pr will be expected in 2.1.0. 

May take a bit longer to be merged, since there are still some pull requests in the 2.0.0 milestone have to be tested or working on it (https://github.com/louislam/uptime-kuma/milestone/24).

adamhancock commented:
Hi @louislam - I have the remote browser minimum functionality working.

Will continue to improve on it and make sure tests are passing. 

<img width="1199" alt="image" src="https://github.com/louislam/uptime-kuma/assets/23136688/0aa86127-b1a8-4c7d-8be4-170f5d077626">

<img width="991" alt="image" src="https://github.com/louislam/uptime-kuma/assets/23136688/5fb22f7e-a66f-41a2-985b-f9a4ba24d7ec">

chakflying commented:
- No need to include `package-lock` since there are no package changes.
- I think the term "local" can be confusing, since there are 2 browsers involved: the one the user is using, and the one run by the server. Open to suggestions, but maybe the option should be named "On server"?

adamhancock commented:
> * No need to include `package-lock` since there are no package changes.

Good catch, reverted it. 

> * I think the term "local" can be confusing, since there are 2 browsers involved: the one the user is using, and the one run by the server. Open to suggestions, but maybe the option should be named "On server"?

You're right. I've updated it to be a checkbox to toggle remote browser to make it less confusing. 
![browserstoggle](https://github.com/louislam/uptime-kuma/assets/23136688/f651674a-ab32-4972-8b4d-ed818f4393bf)

louislam commented:
There are some 2.0.0 pr is stalled, I swap your pr with them.

inkly commented:
Hello Adam, it's possible to check with all remote browser?

louislam commented on src/lang/en.json:
```diff
@@ -859,5 +859,13 @@
+    "Remote Browser": "Remote Browser",
+    "Add a Remote Browser": "Add a Remote Browser",
+    "Remote Browser not found!": "Remote Browser not found!",
+    "remoteBrowsersDescription": "Remote Browsers are an alternative to running Chromium locally. Setup with a service like browserless.io or connect to your self-hosted container: https://hub.docker.com/r/browserless/chrome",
```
`https://hub.docker.com/r/browserless/chrome` is no need to be here. It can be convert to `{0}` and also should be clickable.

louislam commented on src/components/RemoteBrowserDialog.vue:
```diff
@@ -0,0 +1,185 @@
+                            <div class="form-text mt-3">
+                                {{ $t("Examples") }}:
+                                <ul>
+                                    <li>wss://chrome.browserless.io/playwright?token=YOUR-API-TOKEN</li>
```
`ws://` should be better. I think non-developers may not be faimilar with the WebSocket protocol and I think people will likely to test it without TLS first.

louislam commented:
Tested it. Didn't find any big issues. It is a solid pull request.

louislam commented on src/lang/en.json:
```diff
@@ -863,7 +863,7 @@
     "Remote Browser": "Remote Browser",
     "Add a Remote Browser": "Add a Remote Browser",
     "Remote Browser not found!": "Remote Browser not found!",
-    "remoteBrowsersDescription": "Remote Browsers are an alternative to running Chromium locally. Setup with a service like browserless.io or connect to your self-hosted container: https://hub.docker.com/r/browserless/chrome",
```
I think my last comment may be not clear, the url is good to have, but it should be in <a></a> and no need to be translatable.

CommanderStorm replied:
Here are the docs for component interpolation (which is how you can integrate links into translation without resoring to `v-html`): https://vue-i18n.intlify.dev/guide/advanced/component

louislam replied:
@CommanderStorm Actually we don't need this, because translating a url is meaningless. What I meant is that we just need to put the raw url in the vue component, no need to be in i18n.

(Off topic, I just sent a email to you hope you can receive it)

CommanderStorm replied:
@louislam We are talking past each other, component translation or utilising `v-html` (not recommended, this would allow our translators to XSS us) is necessary because of vue's html-escaping.

I did not suggest translating the url ^^

example (not using the slot syntax, unshure which one is better):

- vue:
  ```
  <i18n-t keypath="remoteBrowsersDescription">
    <a href="https://browserless.io" target="_blank">browserless.io</a>
    <a href="https://hub.docker.com/r/browserless/chrome" target="_blank">{{ $t('self-hosted container') }}</a>
  </i18n-t>
  ```
- `en.json`:
  ```diff
  - "remoteBrowsersDescription": "Remote Browsers are an alternative to running Chromium locally. Setup with a service like browserless.io or connect to your self-hosted container: https://hub.docker.com/r/browserless/chrome",
  + "remoteBrowsersDescription": "Remote Browsers are an alternative to running Chromium locally. Setup with a service like {0} or connect to your {0}",
  + "self-hosted container": "self-hosted container",
  ```

chakflying replied:
> utilising v-html (not recommended, this would allow our translators to XSS us

Wait is this actually possible? There are currently a few places where we do use `v-html` for translation 🤔

louislam replied:
> Wait is this actually possible? 

Yes, it should accept `<script></script>`. Bad translators should be able to add it via Weblate, but usually I will have a quick look before merge those translation PRs (#3891). So it won't be a big issue I think.

adamhancock replied:
I've updated the translation to the use the same as @CommanderStorm's example in en.json

Stitch10925 commented:
Pardon my ignorance, but what would be a use-case for remote browsers?
